File: ios/Memora/MemoraViewModel.swift
import SwiftUI

private let sampleOCR = "MILK\nMFG 09/25\nBATCH 84921\nEXP 08/27"

@MainActor
final class MemoraViewModel: ObservableObject {

    @Published private(set) var memories: [Memory] = MemoryStore.load()
    @Published var sourceText: String = sampleOCR
    @Published var candidate: MemoryCandidate = MemoryExtractor.extract(sampleOCR)
    @Published var image: UIImage?
    @Published var dateError = false

    var activeMemories: [Memory] {
        memories
            .filter { MemoryStatusResolver.status(for: $0) != .completed }
            .sorted { lhs, rhs in
                let lu = MemoryStatusResolver.status(for: lhs).urgency
                let ru = MemoryStatusResolver.status(for: rhs).urgency
                return lu == ru ? lhs.dateISO < rhs.dateISO : lu < ru
            }
    }

    var completedMemories: [Memory] {
        memories.filter { MemoryStatusResolver.status(for: $0) == .completed }
    }

    var summary: (overdue: Int, dueToday: Int, upcoming: Int) {
        var overdue = 0, dueToday = 0, upcoming = 0
        for memory in activeMemories {
            switch MemoryStatusResolver.status(for: memory) {
            case .overdue: overdue += 1
            case .dueToday: dueToday += 1
            case .upcoming: upcoming += 1
            case .completed: break
            }
        }
        return (overdue, dueToday, upcoming)
    }

    var candidateStatus: MemoryStatus {
        MemoryStatusResolver.isValidISODate(candidate.dateISO)
            ? MemoryStatusResolver.status(forDate: candidate.dateISO)
            : .upcoming
    }

    func onImageSelected(_ image: UIImage?) {
        self.image = image
        candidate = MemoryExtractor.extract(sourceText)
    }

    func analyze() {
        candidate = MemoryExtractor.extract(sourceText)
    }

    func updateDate(_ value: String) {
        candidate.dateISO = value
        dateError = false
    }

    func confirm() {
        guard MemoryStatusResolver.isValidISODate(candidate.dateISO) else {
            dateError = true
            return
        }

        let notificationId = ReminderScheduler.schedule(candidate)
        let memory = Memory(
            id: String(Int(Date().timeIntervalSince1970 * 1000)),
            title: candidate.title,
            eventType: candidate.eventType,
            dateISO: candidate.dateISO,
            confidence: candidate.confidence,
            sourceText: candidate.sourceText,
            createdAtISO: ISO8601DateFormatter().string(from: Date()),
            completedAtISO: nil,
            notificationId: notificationId
        )
        memories.insert(memory, at: 0)
        persist()
        resetCapture()
    }

    func complete(_ id: String) {
        guard let index = memories.firstIndex(where: { $0.id == id }) else { return }
        if let notificationId = memories[index].notificationId {
            ReminderScheduler.cancel(notificationId)
        }
        memories[index].completedAtISO = ISO8601DateFormatter().string(from: Date())
        persist()
    }

    func delete(_ id: String) {
        guard let memory = memories.first(where: { $0.id == id }) else { return }
        if let notificationId = memory.notificationId {
            ReminderScheduler.cancel(notificationId)
        }
        memories.removeAll { $0.id == id }
        persist()
    }

    private func resetCapture() {
        sourceText = sampleOCR
        candidate = MemoryExtractor.extract(sampleOCR)
        image = nil
    }

    private func persist() {
        MemoryStore.save(memories)
    }
}

File: android/app/src/main/java/com/meruvakirankumar/memora/ui/MemoraViewModel.kt
package com.meruvakirankumar.memora.ui

import android.app.Application
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.ImageBitmap
import androidx.lifecycle.AndroidViewModel
import com.meruvakirankumar.memora.data.MemoryRepository
import com.meruvakirankumar.memora.logic.MemoryExtractor
import com.meruvakirankumar.memora.logic.MemoryStatusResolver
import com.meruvakirankumar.memora.model.EventType
import com.meruvakirankumar.memora.model.Memory
import com.meruvakirankumar.memora.model.MemoryCandidate
import com.meruvakirankumar.memora.model.MemoryStatus
import com.meruvakirankumar.memora.notifications.ReminderScheduler
import java.time.Instant

private const val SAMPLE_OCR = "MILK\nMFG 09/25\nBATCH 84921\nEXP 08/27"

class MemoraViewModel(application: Application) : AndroidViewModel(application) {

    private val repository = MemoryRepository(application)
    private val reminderScheduler = ReminderScheduler(application)

    private val memories = mutableStateListOf<Memory>()

    var sourceText by mutableStateOf(SAMPLE_OCR)
        private set
    var candidate by mutableStateOf(MemoryExtractor.extract(SAMPLE_OCR))
        private set
    var imageBitmap by mutableStateOf<ImageBitmap?>(null)
        private set
    var dateError by mutableStateOf(false)
        private set

    init {
        memories.addAll(repository.load())
    }

    val activeMemories: List<Memory>
        get() = memories
            .filter { MemoryStatusResolver.statusFor(it) != MemoryStatus.COMPLETED }
            .sortedWith(
                compareBy(
                    { MemoryStatusResolver.urgency(MemoryStatusResolver.statusFor(it)) },
                    { it.dateIso },
                ),
            )

    val completedMemories: List<Memory>
        get() = memories.filter { MemoryStatusResolver.statusFor(it) == MemoryStatus.COMPLETED }

    fun summary(): Triple<Int, Int, Int> {
        var overdue = 0
        var dueToday = 0
        var upcoming = 0
        activeMemories.forEach {
            when (MemoryStatusResolver.statusFor(it)) {
                MemoryStatus.OVERDUE -> overdue++
                MemoryStatus.DUE_TODAY -> dueToday++
                MemoryStatus.UPCOMING -> upcoming++
                MemoryStatus.COMPLETED -> Unit
            }
        }
        return Triple(overdue, dueToday, upcoming)
    }

    fun onSourceTextChanged(value: String) {
        sourceText = value
    }

    fun onImageSelected(bitmap: ImageBitmap?) {
        imageBitmap = bitmap
        candidate = MemoryExtractor.extract(sourceText)
    }

    fun analyze() {
        candidate = MemoryExtractor.extract(sourceText)
    }

    fun updateTitle(value: String) {
        candidate = candidate.copy(title = value)
    }

    fun updateEventType(value: EventType) {
        candidate = candidate.copy(eventType = value)
    }

    fun updateDate(value: String) {
        candidate = candidate.copy(dateIso = value)
        dateError = false
    }

    fun candidateStatus(): MemoryStatus =
        if (MemoryStatusResolver.isValidIsoDate(candidate.dateIso)) {
            MemoryStatusResolver.statusForDate(candidate.dateIso)
        } else {
            MemoryStatus.UPCOMING
        }

    fun confirm(): Boolean {
        if (!MemoryStatusResolver.isValidIsoDate(candidate.dateIso)) {
            dateError = true
            return false
        }

        val notificationId = reminderScheduler.schedule(candidate)
        val memory = Memory(
            id = System.currentTimeMillis().toString(),
            title = candidate.title,
            eventType = candidate.eventType,
            dateIso = candidate.dateIso,
            confidence = candidate.confidence,
            sourceText = candidate.sourceText,
            createdAtIso = Instant.now().toString(),
            notificationId = notificationId,
        )
        memories.add(0, memory)
        persist()
        resetCapture()
        return true
    }

    fun complete(memoryId: String) {
        val index = memories.indexOfFirst { it.id == memoryId }
        if (index < 0) return
        val target = memories[index]
        target.notificationId?.let { reminderScheduler.cancel(it) }
        memories[index] = target.copy(completedAtIso = Instant.now().toString())
        persist()
    }

    fun delete(memoryId: String) {
        val target = memories.firstOrNull { it.id == memoryId } ?: return
        target.notificationId?.let { reminderScheduler.cancel(it) }
        memories.remove(target)
        persist()
    }

    private fun resetCapture() {
        sourceText = SAMPLE_OCR
        candidate = MemoryExtractor.extract(SAMPLE_OCR)
        imageBitmap = null
    }

    private fun persist() {
        repository.save(memories.toList())
    }
}
